daemon/internal/cloud_client/change_active_traces_event.go

```diff
@@ -0,0 +1,77 @@
+package cloud_client
+
+import (
+	"github.com/akitasoftware/akita-cli/printer"
+	"github.com/akitasoftware/akita-libs/akid"
+	"github.com/akitasoftware/akita-libs/daemon"
+)
+
+// Indicates that the set of active traces for a service has changed. This is
+// added to the cloud client's main event channel when we get responses to
+// longPollServiceEvent.
+type changedActiveTracesEvent struct {
+	// The service for which the set of active traces has changed.
+	serviceID akid.ServiceID
+
+	// Specifies which traces have been activated, and which have been
+	// deactivated.
+	activeTraceDiff daemon.ActiveTraceDiff
+}
+
+func newChangeActiveTracesEvent(serviceID akid.ServiceID, activeTraceDiff daemon.ActiveTraceDiff) changedActiveTracesEvent {
+	return changedActiveTracesEvent{
+		serviceID:       serviceID,
+		activeTraceDiff: activeTraceDiff,
+	}
+}
+
+func (event changedActiveTracesEvent) handle(client *cloudClient) {
+	printer.Debugf("Handling changed-traces event for service %s\n", akid.String(event.serviceID))
+
+	serviceInfo, ok := client.serviceInfoByID[event.serviceID]
+	if !ok {
+		printer.Debugf("Ignoring diff for unknown service: %s\n", akid.String(event.serviceID))
```

Comment:
If this shouldn't ever happen, then perhaps the log message should be Warning.

Reply:
I've put most of the logging in Debug because I'm not sure how useful these messages would be to our users. Would they understand what this meant if they saw this as a warning? What could they do to address the warning?

Reply:
My goal is that if the user has a problem, we have some chance of diagnosing it from the existing logs alone. This means putting errors in Warning or Info, even if we don't expect the user to be able to do anything about it, simply to avoid "can you turn debug on and make it happen again."  The only reason not to run in Debug mode all the time is space constraints. ​

A secondary goal is that when there *is* a problem, the user has some chance of noticing and contacting us about it, which they can't do if they're silenced by default. While it would be better to have end-to-end telemetry that either we or the user could monitor, "hey, my log is getting spammed by X" does happen.

Because we don't actually have an alerting mechanism (nobody is getting paged or emailed when things go really wrong), I think we should err on the side of putting more into the log as it's low-cost and serves both these goals.

Reply:
Done in 9eca8de.